scheds/rust/scx_rusty/src/bpf/main.bpf.c

```diff
@@ -904,18 +998,46 @@ void BPF_STRUCT_OPS(rusty_running, struct task_struct *p)
 	 */
 	if (vtime_before(domc->vtime_now, p->scx.dsq_vtime))
 		domc->vtime_now = p->scx.dsq_vtime;
+
+	if (vtime_before(domc->vtime_max, p->scx.dsq_vtime)) {
+		domc->vtime_max = p->scx.dsq_vtime;
+		domc->vtime_max_cpu = scx_bpf_task_cpu(p);
+	}
 }
 
 void BPF_STRUCT_OPS(rusty_stopping, struct task_struct *p, bool runnable)
 {
 	struct task_ctx *taskc;
+	struct dom_ctx *domc;
+	u64 *cpu_vtime;
+	u32 dom_id;
+	s32 cpu;
 
 	if (fifo_sched)
 		return;
 
+	cpu = scx_bpf_task_cpu(p);
+	cpu_vtime = MEMBER_VPTR(cpu_vtimes, [cpu]);
+
+	if (preempted_cpumask)
+		bpf_cpumask_clear_cpu(cpu, preempted_cpumask);
+
+	if (cpu_vtime)
+		*cpu_vtime = 0;
+
 	if (!(taskc = lookup_task_ctx(p)))
 		return;
 
+	dom_id = taskc->dom_id;
+	domc = bpf_map_lookup_elem(&dom_data, &dom_id);
+	if (!domc) {
+		scx_bpf_error("Failed to lookup dom[%u]", dom_id);
+		return;
+	}
+
+	if (domc->vtime_max == p->scx.dsq_vtime)
+		set_next_dom_vtime_max(domc, dom_id, cpu);
```

Comment:
I'm not sure I understand this conditional check. we're updating the vtime max for the domain if this task was previously the task with the highest vtime? is this just so that we only do it on a single core?

Reply:
This might be unnecessary, basically if the task is "stopping" then it's no longer the task with the highest vtime for the domain, so we need to pick the next highest so we know what can be preempted in this domain.

Reply:
Ah, I see. So, stopping doesn't necessarily imply no longer runnable. It just means that the core is temporarily not running the task anymore. You'll also want to check `!runnable` here, and I wonder if it might be slightly less confusing to do `domc->vtime_max_cpu == cpu` (not sure it matters though)?

I'm not sure this is worth doing though. It seems potentially undesirable to do a # CPU for loop every time a task stops running on the previously-max-vtime cpu. Consider that this is not going to be an uncommon thing -- it should happen roughly every 1/#CPUs in domain rescheds. A few thoughts:

1. This seems like something that's probably better to start doing only when util >= some threshold, so we don't tax the under-utilized case unnecessarily
2. It doesn't have to be in the first version of this code that lands, but I wonder if we could somehow use an in-BPF rbtree to do this? that way it would be an O(log(# domain CPUs)) operation to insert and remove running tasks, so we wouldn't have to do this O(n) lookup and "rebalance". The only drawback here is we'd be doing tree operations on (almost) every single running / stopping call. hmm, not sure. something to think about

Reply:
> So, stopping doesn't necessarily imply no longer runnable. It just means that the core is temporarily not running the task anymore.

Just for clarification - if the cpu is not running the task (with the highest vtime) then doesn't that mean, in this context, that we don't want kick this cpu in the preemption path in `select_cpu`?

> This seems like something that's probably better to start doing only when util >= some threshold, so we don't tax the under-utilized case unnecessarily

Sorry, what do you mean by "util" here?